src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java

```diff
@@ -0,0 +1,256 @@
+            if (candidate != null) {
+                candidate.rebuildNetwork();
+                return;
+            }
+        }
+    }
+
+    @Override
+    public void writeInitialSyncData(PacketBuffer buf) {
+        super.writeInitialSyncData(buf);
+        buf.writeBoolean(controllerPos != null);
+        if (controllerPos != null) {
+            buf.writeBlockPos(controllerPos);
+        }
+    }
+
+    @Override
+    public void receiveInitialSyncData(PacketBuffer buf) {
+        super.receiveInitialSyncData(buf);
+        if (buf.readBoolean()) {
+            controllerPos = buf.readBlockPos();
+            scheduleRenderUpdate();
+        }
+    }
+
+    @Override
+    public void receiveCustomData(int dataId, PacketBuffer buf) {
+        super.receiveCustomData(dataId, buf);
+        if (dataId == GregtechDataCodes.UPDATE_CONTROLLER_POS) {
+            this.controllerPos = buf.readBlockPos();
+            this.controller.clear();
+
+            if (this.connectedIcon != null) {
+                this.connectedIcon.setButtonTexture(GuiTextures.GREGTECH_LOGO);
+                this.connectedIcon.setTooltipText("gregtech.machine.quantum_storage.connected",
+                        controllerPos.getX(), controllerPos.getZ(), controllerPos.getY());
+            }
+            scheduleRenderUpdate();
+        } else if (dataId == GregtechDataCodes.REMOVE_CONTROLLER) {
+            this.controllerPos = null;
+            this.controller.clear();
+            if (this.connectedIcon != null) {
+                this.connectedIcon.setButtonTexture(GuiTextures.GREGTECH_LOGO_DARK);
+                this.connectedIcon.setTooltipText(null);
+            }
+            scheduleRenderUpdate();
+        } else if (dataId == GregtechDataCodes.LOCATE_CONTROLLER && buf.readBoolean()) {
+            // tell controller to highlight
+            BlockPosHighlightRenderer.renderBlockBoxHighLight(getControllerPos(), 6000, 1500);
+            Minecraft.getMinecraft().player.closeScreen();
+        }
+    }
+
+    @Override
+    public NBTTagCompound writeToNBT(NBTTagCompound data) {
+        NBTTagCompound tagCompound = super.writeToNBT(data);
+        tagCompound.setBoolean("HasController", controllerPos != null);
+        if (controllerPos != null) {
+            tagCompound.setLong("ControllerPos", controllerPos.toLong());
+        }
+        return tagCompound;
+    }
+
+    @Override
+    public void readFromNBT(NBTTagCompound data) {
+        super.readFromNBT(data);
+        if (data.getBoolean("HasController")) {
+            this.controllerPos = BlockPos.fromLong(data.getLong("ControllerPos"));
+        }
+    }
+
+    protected ClickButtonWidget createConnectedGui(int y) {
+        connectedIcon = new ClickButtonWidget(151, y, 18, 18, "",
+                clickData -> {
+                    writeCustomData(GregtechDataCodes.LOCATE_CONTROLLER, buffer -> {
+                        buffer.writeBoolean(this.isConnected());
+                    });
```

Comment:
Instead of writing if the quantum part is connected in the custom data, and then always writing the custom data, why not just check if the quantum part is connected first, and then only write the packet if it is?

Reply:
there would need to be a packet for when the quantum part is disconnected, to update the state of the icon

Reply:
nvm, i understand what you mean now